flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractComponentDataGenerator.java

```diff
@@ -126,6 +129,20 @@ protected void registerRenderedComponent(String itemKey,
 
         Element element = component.getElement();
         getContainer().appendChild(element);
+        NodeOwner owner = getContainer().getNode().getOwner();
+        UI containerUi = null;
+        if (owner instanceof StateTree) {
+            containerUi = ((StateTree) owner).getUI();
+        }
+        if (containerUi != null && component.getUI().isPresent()
+                && containerUi != component.getUI().get()) {
+            throw new IllegalStateException("The component '"
+                    + component.getClass()
+                    + "' is already attached to a UI instance which differs "
+                    + "from the conainer's UI instance. It means that the component instance is "
+                    + "reused instead being produced every time on 'createComponent' call."
+                    + " Check whether the component instance is a singleton or has inapropriate Spring scope.");
```

Comment:
```suggestion
                    + " Check whether the component instance is a singleton or has inappropriate Spring scope.");
```
Should we say `DI container` instead of Spring as you could use the wrong CDI scope also.